### PR TITLE
Simplify implementation of GC BIFs

### PR DESCRIPTION
**Pull request by bjorng**
This pull request simplifies the implementation of the "GC BIFs" so that they no longer need to do a garbage collection, removing duplicate code for all GC BIFs in the runtime system, as well as reducing the size of the loaded BEAM code by using shorter instructions calling those BIFs.

See the commit messages for more details.

This pull request intentionally does not update the compiler. Emitting `bif` instructions instead of `gc_bif` instructions does not seem improve the code quality. However, by adding some additional optimizations, it seems that some register shuffling code could be avoided by using plain `bif` instructions, but that will be subject of a future investigation and possible pull request.

### CODE REVIEW

File: erts/emulator/beam/bif.tab [hunk 1]
```diff
@@ -623,7 +623,7 @@ bif io:printable_range/0
 bif re:inspect/2
 
 ubif erlang:is_map/1
-gcbif erlang:map_size/1
+ubif erlang:map_size/1
```
Comment (michalmuskala) [inline erts/emulator/beam/bif.tab]: Does that call the `map_size_1` function in `erl_map.c`? If so, it calls `HAlloc` and that's supposedly not allowed now, isn't it?

Comment (bjorng) [reply erts/emulator/beam/bif.tab]: Good catch.

Yes, calling `HAlloc()` is not allowed and if `hsz` (the size argument passed to `HAlloc()` would ever be nonzero, the runtime system would crash.

In practice, it is not possible to create a map whose size can't be represented in a small. So I have pushed a commit that removes the call to `HAlloc()` and the superfluous calls to `erts_bld_uint()`.

File: erts/emulator/beam/arith_instrs.tab [hunk 2]
```diff
@@ -320,7 +307,9 @@ shift.execute(Fail, Live, Dst) {
         }
```
Comment (garazdawi) [inline erts/emulator/beam/arith_instrs.tab]: Should we move the entire big_shift logic out of the emulator and into big.c?

File: erts/emulator/beam/bif.tab (hunk 1 above)
Comment (garazdawi) [reply erts/emulator/beam/bif.tab]: Would it be possible to set some kind of debug flag that triggers an assert if HAlloc is called in an incorrect function?

File: erts/emulator/beam/arith_instrs.tab [hunk 3]
```diff
@@ -19,21 +19,15 @@
...
-OUTLINED_ARITH_2(Fail, Live, Name, BIF, Op1, Op2, Dst) {
+OUTLINED_ARITH_2(Fail, Name, BIF, Op1, Op2, Dst) {
     Eterm result;
-    Uint live = $Live;
-    HEAVY_SWAPOUT;
-    reg[live] = $Op1;
-    reg[live+1] = $Op2;
-    result = erts_gc_$Name (c_p, reg, live);
-    HEAVY_SWAPIN;
+    result = erts_$Name (c_p, $Op1, $Op2);
```
Comment (garazdawi) [inline erts/emulator/beam/arith_instrs.tab]: Add check here that stop and htop have not moved?

File: erts/emulator/beam/bif_instrs.tab [hunk 4]
```diff
@@ -37,7 +37,9 @@ CALL_GUARD_BIF(BF, TmpReg, Dst) {
     PROCESS_MAIN_CHK_LOCKS(c_p);
     ASSERT(!ERTS_PROC_IS_EXITING(c_p));
     ERTS_CHK_MBUF_SZ(c_p);
+    DEBUG_SWAPOUT;
     result = (*$BF)(c_p, $TmpReg, I);
+    DEBUG_SWAPIN;
```
Comment (garazdawi) [inline erts/emulator/beam/bif_instrs.tab]: Again, maybe check that stop and htop have not moved?

File: erts/emulator/beam/erl_bif_guard.c [hunk 5]
```diff
@@ -19,7 +19,12 @@
  */
 
 /*
- * Numeric guard BIFs.
+ * This file implements the former GC BIFs. They used to do a GC when
+ * they needed heap space. Because of changes to the implementation of
+ * literals, those BIFs are now allowed to allocate heap fragments
+ * (using HeapFragOnlyAlloc()). Not that they must NOT call HAlloc(),
```
Comment (garazdawi) [inline erts/emulator/beam/erl_bif_guard.c]: "Note that they must NOT" (missing e from Note)

File: erts/emulator/beam/erl_map.c [hunk 6]
```diff
@@ -125,15 +125,18 @@ BIF_RETTYPE map_size_1(BIF_ALIST_1) {
...
-	(void) erts_bld_uint(NULL, &hsz, size);
-	hp = HAlloc(BIF_P, hsz);
-	res = erts_bld_uint(&hp, NULL, size);
-	BIF_RET(res);
+
+        /*
+         * In practice, the size of a map will always fit in
+         * a small.
+         */
+        ASSERT(IS_USMALL(0, size));
```
Comment (garazdawi) [inline erts/emulator/beam/erl_map.c]: Use ERTS_ASSERT?

File: erts/emulator/beam/bif.tab (hunk 1 above)
Comment (bjorng) [reply erts/emulator/beam/bif.tab]: I have added assertions that htop and stop are not changed. That should catch any calls to `HAlloc()` that allocate anything. Having a separate `HAllocForbidden` flag could potentially catch calls `HAlloct()` that don't actually allocate anything, but is seems to be overkill.

File: erts/emulator/beam/arith_instrs.tab (hunk 2 above)
Comment (bjorng) [reply erts/emulator/beam/arith_instrs.tab]: I would want to consider that question separately from this pull request.

File: erts/emulator/beam/bif_instrs.tab (hunk 4 above)
Comment (bjorng) [reply erts/emulator/beam/bif_instrs.tab]: Yes. Done.

File: erts/emulator/beam/erl_bif_guard.c (hunk 5 above)
Comment (bjorng) [reply erts/emulator/beam/erl_bif_guard.c]: Fixed.

File: erts/emulator/beam/erl_map.c (hunk 6 above)
Comment (bjorng) [reply erts/emulator/beam/erl_map.c]: I have reworded the comment and explained the purpose of the ASSERT (which is to catch a problem if we would ever reduce the number of useful bits in small).

**Pull request merged**